launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,457 @@
+                    '"fixture" keyword argument is required in a pytest.mark.launch_testing() '
+                    f'decorator: \n{get_error_context_from_obj(item.obj)}'
+                )
+                return None
+            fixture = marker.kwargs['fixture']
+            fixturename = fixture.__name__
+            scope = fixture._pytestfixturefunction.scope
+            is_shutdown = marker.kwargs.get('shutdown', False)
+            items = generate_test_items(collector, name, obj, fixturename, is_shutdown)
+            if need_shutdown_test_item(obj) and scope != 'function':
+                # for function scope we only need one shutdown item
+                # if not we're going to use two event loops!!!
+                shutdown_items = generate_test_items(collector, name, obj, fixturename, True)
+                for item, shutdown_item in zip(items, shutdown_items):
+                    item._launch_testing_shutdown_item = shutdown_item
+                items.extend(shutdown_items)
+            return items
+
+
+def get_error_context_from_obj(obj):
+    """Return formatted information of the object location."""
+    try:
+        fspath = inspect.getsourcefile(obj)
+    except TypeError:
+        return 'location information of the object not found'
+    try:
+        lines, lineno = inspect.getsourcelines(obj)
+    except IOError:
+        return f'file {fspath}: source code not available'
+    error_msg = f'file {fspath}, line {lineno}'
+    for line in lines:
+        line = line.rstrip()
+        error_msg += f'\n  {line}'
+        if line.lstrip().startswith('def'):
+            break
+    return error_msg
+
+
+@pytest.mark.trylast
+def pytest_collection_modifyitems(session, config, items):
+    """Reorder tests, so shutdown tests happen after the corresponding fixture teardown."""
+
+    def cmp(left, right):
+        leftm = left.get_closest_marker('launch_testing')
+        rightm = right.get_closest_marker('launch_testing')
+        if None in (leftm, rightm):
+            return 0
+        fixture = leftm.kwargs['fixture']
+        if fixture is not rightm.kwargs['fixture']:
+            return 0
+        if fixture._pytestfixturefunction.scope == 'function':
+            return 0
+        left_is_shutdown = int(left._launch_testing_is_shutdown)
+        right_is_shutdown = int(right._launch_testing_is_shutdown)
+        return left_is_shutdown - right_is_shutdown
+
+    # python sort is guaranteed to be stable
+    items.sort(key=functools.cmp_to_key(cmp))
+    # This approach doesn't work, because we cannot guarantee when fixture
+    # finalizers are going to be run.
+    # remove the launch_testing fixture from shutdown tests
+    # so launch service shutdown happens right before... by magic!!!
+    for item in items:
+        marker = item.get_closest_marker('launch_testing')
+        if marker is not None and item._launch_testing_is_shutdown:
+            item.fixturenames.remove(marker.kwargs['fixture'].__name__)
```

Comment:
Wild idea: what if we use dependent fixtures to handle launch service lifecycle instead of finalizers?

Let's say we have a launch fixture A. On decoration, we introduce two more fixtures B and C with the same scope. Both B and C depend on A. B takes the value of A and manages the launch service lifecycle: it pushes it into the event loop, yields, and on finalization it shuts down (or not) and waits for it to wrap up. C takes the value of A, runs the launch service to completion, and returns. Regular launch tests depend on B. Post-shutdown launch tests depend on A, unless there's only post-shutdown launch tests in which case they all depend on C. 

That should take care of test ordering, passing extra arguments, and the _all shutdown tests_ scenario. WDYT?

Reply:
> @ivanpauno hmm I have the impression this won't work when all tests are shutdown tests.

Yes, this is still true.
I'm inclined to delete that option.

> Wild idea: what if we use dependent fixtures to handle launch service lifecycle instead of finalizers?

Maybe that works, I have to another look to this alternative.

Reply:
Yeah, it may or may not. When I thought about it looked like it did :sweat_smile:

Reply:
> Yeah, it may or may not. When I thought about it looked like it did sweat_smile

I gave it a try, it's impossible to support parameterized fixtures in that way.
(maybe it's possible, but pytest doesn't seem to have been think to do that kind of trick)

Reply:
we discussed this offline, we agreed the current approach seems fine.